src/pykiso/tool/testrail/api.py

```diff
@@ -0,0 +1,394 @@
+        headers["Content-Type"] = content_type
+        return headers
+
+    @classmethod
+    def get_projects(cls, base_url: str, user: str, password: str) -> dict:
+        """Retrieve all available projects under TestRail
+
+        :param base_url: TestRail's base url
+        :param user: user's session id
+        :param password: user's password or API key
+
+        :return: all available projects information
+        """
+        url = cls._construct_url(base_url, cls.API_VERSION, uri="get_projects")
+        headers = cls._construct_header(user, password, ContentType.JSON.value)
+        query_result = HttpRequest.get(url=url, headers=headers)
+        return json.loads(query_result.content)
+
+    @classmethod
+    def get_suites(
+        cls, base_url: str, user: str, password: str, project_id: int
+    ) -> dict:
+        """Retrieve all available suite under the given project.
+
+        :param base_url: TestRail's base url
+        :param user: user's session id
+        :param password: user's password
+        :param project_id: project id in which to search for suites
+
+        :return: all available suites information
+        """
+        url = cls._construct_url(
+            base_url, cls.API_VERSION, uri=f"get_suites/{project_id}"
+        )
+        headers = cls._construct_header(user, password, ContentType.JSON.value)
+        query_result = HttpRequest.get(url=url, headers=headers)
+        return json.loads(query_result.content)
+
+    @classmethod
+    def get_cases(
+        cls, base_url: str, user: str, password: str, project_id: int, suite_id: int
+    ) -> dict:
+        """Retrieve all available test cases under the given project and
+        suite.
+
+        :param base_url: TestRail's base url
+        :param user: user's session id
+        :param password: user's password
+        :param project_id: targeted project id
+        :param suite_id: targeted suite id
+
+        :return: all available runs information
+        """
+        url = cls._construct_url(
+            base_url, cls.API_VERSION, uri=f"get_cases/{project_id}&suite_id={suite_id}"
+        )
+        headers = cls._construct_header(user, password, ContentType.JSON.value)
+        query_result = HttpRequest.get(url=url, headers=headers)
+        return json.loads(query_result.content)
+
+    @classmethod
+    def async_get_cases(
+        cls,
+        base_url: str,
+        user: str,
+        password: str,
+        project_id: int,
+        suite_ids: List[int],
+    ) -> List[dict]:
+        """Asynchronously get test cases information.
+
+        :param base_url: TestRail's base url
+        :param user: user's session id
+        :param password: user's password
+        :param project_id: TestRail's project id
+        :param suite_ids: all suite ids to search for test cases
+
+        :return: all test case information
+        """
+        with concurrent.futures.ThreadPoolExecutor(
```

Comment:
Why not using asyncio? (https://docs.python.org/3/library/asyncio-protocol.html#connecting-existing-sockets)

Reply:
An [Executor](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor) subclass that uses a pool of at most max_workers threads to execute calls asynchronously.

Reply:
Got it but my question was more towards the differences in the technology.

Reply:
Find below a little list of difference:

ThreadPoolExecutor

    Uses Threads, not Coroutines.
    System-level, not software-level.
    Preemptive multitasking, not Cooperative.
    Tasks are any function, not constrained.
    Requires Thread Safety.
    Create 100s of Threads, not 1000s

AsyncIO

    Uses Coroutines, not Threads.
    Software-level, not System-level.
    Cooperative multitasking, not Preemptive.
    Awaitable Tasks, not any function.
    No Thread Safety Concerns.
    Create 100,000+ Coroutines, not reasonably limited.